napari/utils/colormaps/categorical_colormap_utils.py

```diff
@@ -0,0 +1,26 @@
+from dataclasses import dataclass
+from itertools import cycle
+
+import numpy as np
+
+
+@dataclass
+class ColorCycle:
```

Comment:
I think that is fine. We will need to add an equality check for the `cycle` iterator. Since I don't think there's a real way to know it's state, should it just always return `False`?

Reply:
Good point, the equality check wasn't needed.

However, the current issue I have is that I am not sure how to allow for JSON serialization via `ColorCycle.json()`. The issue is that we can't serialize the `cycle` object. The way I got around this is by adding the `ColorCycle` class to the `JSON_ENCODERS` dict with a function that just serializes the `ColorCycle.values` array. This doesn't work when `ColorCycle` is an `EventedModel` because it creates a circular import (`JSON_ENCODERS` needs to import `ColorCycle` and `ColorCycle` needs to import `JSON_ENCODERS`). I couldn't find a straightforward way to always exclude a particular field from the `.json()` method. Any thoughts, @sofroniewn and @tlambert03 ?

Reply:
How about adding a custom encoder for a `cycle` where we cast it to a list? Would that work

Reply:
Unfortunately, a cycle can't really be cast to a list because it doesn't does have an end. This is the reason I created the `ColorCycle` dataclass, which just has a field for the cycle and a field containing the array of colors contained in the cycle (`values` field). My thought was that we can then serialize out the `ColorCycle.values` as a list and then skip the `cycle`. This works pretty easily by making `ColorCycle` a dataclass and then adding a custom json encoder as I've done here.

Alternatively, I think we can look into ways to skip the `cycle` during serialization in pydantic. I looked a bit and there are ways we can pass `exclude` arguments to the serialization method, but that would mean that `ColorManager.json()` wouldn't work with default arguments and I think that is undesirable.

Reply:
> Unfortunately, a cycle can't really be cast to a list because it doesn't does have an end. 

Hmm right, i feel like we often make cycles from a list don't we? Or sets of finite elements. I wonder if we might want to store the original list / number of elements somewhere and then grad that during the serialization?

Reply:
I think we're going to come across many cases like this where a python object doesn't have an obvious serialization.  If we release ourselves from the expectation that _all_ objects are going to be pretty human-readable things when serialized, we can just use pickle and serialize to bytes:

```python
In [1]: from itertools import cycle

In [2]: import pickle

In [3]: my_cycle = cycle([1,2,3])

In [4]: serialized = pickle.dumps(my_cycle)

In [5]: serialized
Out[5]: b'\x80\x04\x95G\x00\x00\x00\x00\x00\x00\x00\x8c\titertools\x94\x8c\x05cycle\x94\x93\x94\x8c\x08builtins\x94\x8c\x04iter\x94\x93\x94]\x94(K\x01K\x02K\x03e\x85\x94R\x94K\x00b\x85\x94R\x94]\x94K\x00\x86\x94b.'

In [6]: new_cyle = pickle.loads(serialized)

In [7]: new_cyle
Out[7]: <itertools.cycle at 0x7f9848a609c0>

In [8]: [next(my_cycle) for i in range(8)]
Out[8]: [1, 2, 3, 1, 2, 3, 1, 2]

```

Reply:
> If we release ourselves from the expectation that all objects are going to be pretty human-readable things when serialized, we can just use pickle and serialize to bytes:

I really see this as a last resort, maybe I have a bias against using pickle that comes from some bad interoperability experiences back on some projects at Janelia (tbh though it's more bias/ bad feelings right now then a concrete this is the exact problem, but i can think more about what it was). The further we can get without having to do this the better I think.

Is there any downside to storing a list that we internally convert to a cycle for normal usage (and then delay the pickle decision)? It seems much more transparent to me to just have the list, but I'm wondering if I am missing an application or use case.

Reply:
sure, we can delay it.  the use case is more complicated stuff where we don't want to have to figure out our own custom string representation.  pickle provides a generalizable API with `__getstate__` / `__setstate__`.  There's nothing that magicical about pickle. it just says "give me a function that converts your object to bytes and back ... and I'll call them for you".  It feels very much like json.dumps to me... but for `bytes` instead of strings...  But we can certainly make up our own string representation for things if you'd prefer.

Reply:
> There's nothing that magicical about pickle.

I think the issue from before was cross language compatibility (i.e. outside of python). Possibly not a high priority today, but might be nice one day.

Reply:
yep, good point!

Reply:
> I'd make this an EventedModel model too, and use the Array[float, (-1, 4)] syntax. 

I'm not sure this makes for a good `EventedModel`... this makes more sense as a field.  So it should have 1) a `validate` class method, 2) an `__eq__` method, and 3) after #2297, an `_encode_json` method.  See https://github.com/napari/napari/pull/2297#issuecomment-785079950 for more background on the field vs model distinction.

Reply:
Thank you for the feedback and discussion, @tlambert03 and @sofroniewn ! 

> I'm not sure this makes for a good EventedModel... this makes more sense as a field. So it should have 1) a validate class method, 2) an __eq__ method, and 3) after #2297, an _encode_json method. See #2297 (comment) for more background on the field vs model distinction.

@tlambert03 , I totally agree. This is how the `ColorCycle` is implemented now. It is a dataclass with an `__eq__` method that is a field of the `CategoricalColormap` `EventedModel`. I had previously put the json encoder in the `JSON_ENCODERS` dict, but I see that #2297 is merged, so I will convert to that now!

> Hmm right, i feel like we often make cycles from a list don't we? Or sets of finite elements. I wonder if we might want to store the original list / number of elements somewhere and then grad that during the serialization?

@sofroniewn , this is my intended purpose of the `ColorCycle` object. It has a `values` field, which contains the array elements that are "cycled" in the `cycle object and a `cycle` field, which contains the `cycle` object. Currently, it is implemented as a dataclass (inline with what @tlambert03 described in the comment above (I responded in the paragraph above). My understanding is that you were interested in exploring if the `ColorCycle` object could be an `EventedModel` to help with serialization and event emission, which prompted my question about skipping the `cycle` field for serialization. After playing with it, I agree with @tlambert03  that it makes the most sense to keep `ColorCycle` as a dataclass that is a field of the `CategoricalColormap`. In this way, we can use a `_encode_json` method on the`CategoricalColormap` for serialization and since `CategoricalColormap` is an `EventedModel`, it will hand the event emission when the `ColorCycle` is updated.

> I think the issue from before was cross language compatibility (i.e. outside of python). Possibly not a high priority today, but might be nice one day.

Yes, I agree this is important.

Reply:
Thanks again for the help, @tlambert03 and @sofroniewn ! `ColorCycle` is now a field of `CategoricalColormap` with validate, `__eq__`, and `_encode_json` methods. I like how this ended up!

Reply:
looks good to me too